src/hp/magicsock/conn.rs

```diff
@@ -0,0 +1,3379 @@
+pub struct Options {
+    /// The port to listen on.
+    /// Zero means to pick one automatically.
+    pub port: u16,
+
+    /// Optionally provides a func to be called when endpoints change.
+    #[allow(clippy::type_complexity)]
+    #[debug("on_endpoints: Option<Box<..>>")]
+    pub on_endpoints: Option<Box<dyn Fn(&[cfg::Endpoint]) + Send + Sync + 'static>>,
+
+    /// Optionally provides a func to be called when a connection is made to a DERP server.
+    #[debug("on_derp_active: Option<Box<..>>")]
+    pub on_derp_active: Option<Box<dyn Fn() + Send + Sync + 'static>>,
+
+    /// A callback that provides a `cfg::NetInfo` when discovered network conditions change.
+    #[debug("on_net_info: Option<Box<..>>")]
+    pub on_net_info: Option<Box<dyn Fn(cfg::NetInfo) + Send + Sync + 'static>>,
+    /// Private key for this node.
+    pub private_key: key::node::SecretKey,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Options {
+            port: 0,
+            on_endpoints: None,
+            on_derp_active: None,
+            on_net_info: None,
+            private_key: key::node::SecretKey::generate(),
+        }
+    }
+}
+
+/// Iroh connectivity layer.
+///
+/// This is responsible for routing packets to peers based on peer IDs, it will initially
+/// route packets via a derper relay and transparently try and establish a peer-to-peer
+/// connection and upgrade to it.  It will also keep looking for better connections as the
+/// network details of both endpoints change.
+///
+/// It is usually only necessary to use a single [`Conn`] instance in an application, it
+/// means any QUIC endpoints on top will be sharing as much information about peers as
+/// possible.
+#[derive(Clone, Debug)]
+pub struct Conn {
+    inner: Arc<Inner>,
+    // Empty when closed
+    actor_tasks: Arc<Mutex<Vec<AbortingJoinHandle<()>>>>,
+}
+
+impl Deref for Conn {
+    type Target = Inner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(derive_more::Debug)]
+pub struct Inner {
+    actor_sender: mpsc::Sender<ActorMessage>,
+    /// Sends network messages.
+    network_sender: mpsc::Sender<Vec<quinn_udp::Transmit>>,
+    pub(super) name: String,
+    #[allow(clippy::type_complexity)]
+    #[debug("on_endpoints: Option<Box<..>>")]
+    on_endpoints: Option<Box<dyn Fn(&[cfg::Endpoint]) + Send + Sync + 'static>>,
+    #[debug("on_derp_active: Option<Box<..>>")]
+    pub(super) on_derp_active: Option<Box<dyn Fn() + Send + Sync + 'static>>,
+    /// A callback that provides a `cfg::NetInfo` when discovered network conditions change.
+    #[debug("on_net_info: Option<Box<..>>")]
+    on_net_info: Option<Box<dyn Fn(cfg::NetInfo) + Send + Sync + 'static>>,
+
+    /// Used for receiving DERP messages.
+    network_recv_ch: flume::Receiver<NetworkReadResult>,
+    /// Stores wakers, to be called when derp_recv_ch receives new data.
+    network_recv_wakers: std::sync::Mutex<Option<Waker>>,
```

Comment:
Not a super fan of the explicit waker management. Alternatives are:

- store the sender and receiver in a mutex and then directly call the poll fns on them which allow passing in the context, see https://github.com/n0-computer/quic-rpc/blob/09628eb5a9a7fa7dd51db168723355d4ee420bfd/src/transport/interprocess.rs#L22
- make a thing that encapsulates the waker mgmt.

Now, this code is not wrong or anything. I just usually have a bad feeling with manual waker management code since it is so easy to get subtly wrong.

Reply:
For now I would like to leave it as is, as the last time I tried to simplify it I failed and added more bugs. I would love to avoid this, but I think for the moment it is needed.